core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+  }
+
+  def BIP32DerivationPaths: Vector[BIP32DerivationPath] = {
+    getRecords(BIP32DerivationPathKeyId)
+  }
+
+  def finalizedScriptSigOpt: Option[FinalizedScriptSig] = {
+    getRecords(FinalizedScriptSigKeyId).headOption
+  }
+
+  def finalizedScriptWitnessOpt: Option[FinalizedScriptWitness] = {
+    getRecords(FinalizedScriptWitnessKeyId).headOption
+  }
+
+  def proofOfReservesCommitmentOpt: Option[ProofOfReservesCommitment] = {
+    getRecords(ProofOfReservesCommitmentKeyId).headOption
+  }
+
+  private def getRecords(key: PSBTInputKeyId): Vector[key.RecordType] = {
+    super.getRecords(key, PSBTInputKeyId)
+  }
+
+  def isFinalized: Boolean =
+    getRecords(FinalizedScriptSigKeyId).nonEmpty || getRecords(
+      FinalizedScriptWitnessKeyId).nonEmpty
+
+  /** Finalizes this input if possible, returning None if not */
+  def finalize(input: TransactionInput): Try[InputPSBTMap] = {
+    if (isFinalized) {
+      Success(this)
+    } else {
+      val scriptPubKeyToSatisfyT: Try[ScriptPubKey] = {
+        val witnessUTXOOpt =
+          getRecords(WitnessUTXOKeyId).headOption
+        val nonWitnessUTXOOpt =
+          getRecords(NonWitnessUTXOKeyId).headOption
+
+        (witnessUTXOOpt, nonWitnessUTXOOpt) match {
+          case (None, None) =>
+            Failure(
+              new IllegalStateException(
+                s"Cannot finalize without UTXO record: $this"))
+          case (Some(witnessUtxo), _) =>
+            Success(witnessUtxo.spentWitnessTransaction.scriptPubKey)
+          case (_, Some(utxo)) =>
+            val outputs = utxo.transactionSpent.outputs
+            Try(outputs(input.previousOutput.vout.toInt).scriptPubKey)
+        }
+      }
+
+      scriptPubKeyToSatisfyT.flatMap(finalize)
+    }
+  }
+
+  /** Finalizes this input if possible, returning None if not */
+  private def finalize(spkToSatisfy: ScriptPubKey): Try[InputPSBTMap] = {
+
+    /** Removes non-utxo and non-unkown records, replacing them with finalized records */
+    def wipeAndAdd(
+        scriptSig: ScriptSignature,
+        witnessOpt: Option[ScriptWitness] = None): InputPSBTMap = {
+      val utxos = getRecords(WitnessUTXOKeyId) ++ getRecords(
+        NonWitnessUTXOKeyId)
+      val unknowns =
+        elements.filter(_.isInstanceOf[InputPSBTRecord.Unknown])
+
+      val finalizedScriptSig = FinalizedScriptSig(scriptSig)
+      val recordsToAdd = witnessOpt match {
+        case None => Vector(finalizedScriptSig)
+        case Some(scriptWitness) =>
+          Vector(finalizedScriptSig, FinalizedScriptWitness(scriptWitness))
+      }
+
+      val records = utxos ++ recordsToAdd ++ unknowns
+      InputPSBTMap(records)
+    }
+
+    /** Turns the required PartialSignatures into a ScriptSignature and calls wipeAndAdd
+      * @return None if the requirement is not met
+      */
+    def collectSigs(
+        required: Int,
+        constructScriptSig: Seq[(ECDigitalSignature, ECPublicKey)] => ScriptSignature): Try[
```

Comment:
can't we use the `PartialSignature` type here? `Seq[PartialSignature] => ScriptSignature`